### ISSUE TITLE
MergedMutation

### ISSUE THREAD

**Issue by meh** [async-graphql/async-graphql#539]
## Description of the feature

I'd like to be able to merge mutations (and merge objects with mutations), this would allow me to perform nested mutations.

## Code example

```rust
struct With(Uuid);
struct User(Uuid, String);

#[Object]
impl User {
  fn id(&self) -> ID {
    self.0.into()
  }

  fn name(&self) -> &str {
    &self.1
  }
}

#[Mutation]
impl With {
  fn update(&self) -> Mutation {
    let updated_user = update_user();
    Mutation(self.clone(), updated_user)
  }
}

#[MergedMutation]
struct Mutation(With, User);
```

**Comment by sunli829** [comment]
The only difference between Query and Mutation is the parallel and sequential resolve fields, so I will add a new `serial` attribute to the `SimpleObject` and `Object` macros, which can easily support nested Mutation.

```rust
struct MutationUser(Uuid);

#[Object]
#[graphql(serial)]
impl MutationUser {
    fn update_a(&self) -> bool { ... } 

    fn update_b(&self) -> bool { ... } 
}

struct Mutation;

#[Object]
impl Mutation {
    fn user(id: Uuid) -> MutationUser {
        MutationUser(id)
    }
}
```

execute:

```graphql
mutation {
    user(id: "...") {
        update_a(...)
        update_b(...)
    }
}
```

Is this what you want? @meh

**Comment by meh** [comment]
As long as I can merge them then it's all great.

**Comment by sunli829** [comment]
I don't think I understand the purpose of the example code you gave. Can you explain it in more detail? 😁

**Comment by meh** [comment]
So basically I want to merge a mutation and the object, so that the query can resolve the fields of the object as well as call the mutations.

Otherwise I'd have to reimplement all of the resolvers on the mutation as well.

**Comment by sunli829** [comment]
How do you want to use GraphQL to query this schema?

```rust
struct With(Uuid);
struct User(Uuid, String);

#[Object]
impl User {
  fn id(&self) -> ID {
    self.0.into()
  }

  fn name(&self) -> &str {
    &self.1
  }
}

#[Mutation]
impl With {
  fn update(&self) -> Mutation {
    let updated_user = update_user();
    Mutation(self.clone(), updated_user)
  }
}

#[MergedMutation]
struct Mutation(With, User);
```

**Comment by meh** [comment]
Okay, lemme show you the GraphQL schema I've got and that might make more sense.

```graphql
type Project implements Node {
  # The UUID of the project.
  id: ID!

  # The name of the project.
  name: String

  # The entity the project belongs to.
  owner: Entity!

  # Messages strictly on the project (i.e. not messages on associated
  # objects).
  messages(
    by: MessageFilter! = { id: null, ids: null, label: null, labels: null }
  ): [Message!]!

  # Films within this project.
  films: [ProjectFilm!]!

  # The time the project was created at.
  createdAt: DateTime!

  # The last time the project was updated.
  updatedAt: DateTime

  # The time the project was deleted.
  deletedAt: DateTime
}

type ProjectsMutation {
  create(input: ProjectCreateInput!): Project!
}

type ProjectMutation {
  update(input: ProjectUpdateInput!): Project!
  films: ProjectFilmsMutation!
  film(id: ID!): ProjectFilmMutation!
  labels: ProjectLabelMutation!
}
```

Now in `ProjectsMutation.create` I'd like to return `ProjectMutation`, but I'd also like `ProjectMutation` to have all the resolvers from `Project`.

**Comment by sunli829** [comment]
Can you use `MergedObject` directly?

https://async-graphql.github.io/async-graphql/en/merging_objects.html

**Comment by meh** [comment]
Do you have an example? That also might not work because the mutations carry more information than the object, so I still need to merge them.

If MergedObject also works for mutations then that's great.

**Comment by sunli829** [comment]
Do you want to perform creation and update sequentially like this.
I don't recommend this. GraphQL mutatios are usually not nested, because it is difficult to control their execution order.

```graphql
mutation {
    create(...) {
        update(...)
    }
}
```

**Comment by meh** [comment]
There are other mutations aside from update, but yes. I need these things to be nested to add stuff to the project, and add stuff to the stuff inside project, and so on and so forth.

**Comment by sunli829** [comment]
This is not a good idea. It seems that query and mutation are the same. They can be nested, but in fact, mutations are executed sequentially. This is the only difference between them.
Generally, all mutation operations should be at the top level, and the objects returned by the mutation operations can only be used for queries.

For example:

https://docs.github.com/en/graphql/reference/mutations

**Comment by meh** [comment]
I know they are executed sequentially, why would that be a problem with nesting?

It's mutating nested objects.

**Comment by sunli829** [comment]
Because I need to change the execution behavior of the entire mutation, execute this nested query sequentially.

Previously, the query steps were like this, `b`, `c`, and `d` were resolved in parallel. Now since they are in the mutation, I need to resolved them sequentially.

```graphql
mutation {
    a(...) {         ---------> 1
         b            ---------> 2
         c            ---------> 2
         d            ---------> 2
    }
}
```

```graphql
mutation {
    a(...) {         ---------> 1
         b            ---------> 2
         c            ---------> 3
         d            ---------> 4
    }
}
```

**Comment by sunli829** [comment]
Please see the description of mutation operations in the GraphQL specification.

https://spec.graphql.org/June2018/#sec-Mutation

**Comment by meh** [comment]
Yes, and that's what I want, that's the intended order of operations.

Why would the lack of parallelism be a problem for mutations?

**Comment by sunli829** [comment]
> Yes, and that's what I want, that's the intended order of operations.
> 
> Why would the lack of parallelism be a problem for mutations?

Because only the top level needs to be executed sequentially, and the subsequent ones need to be executed in parallel, if I let the entire mutation be executed sequentially, this does not seem to conform to the GraphQL specification. 😁

**Comment by sunli829** [comment]
Of course, if you really need to do this, I can add an option when creating the schema, which is very easy.

**Comment by meh** [comment]
So wait, nested fields are ALWAYS ran in parallel even if the nested object is a mutation?

I thought the MergedMutation would turn all the resolvers to run sequentially.

**Comment by sunli829** [comment]
> So wait, nested fields are ALWAYS ran in parallel even if the nested object is a mutation?
> 
> I thought the MergedMutation would turn all the resolvers to run sequentially.

In fact, all except the top level are executed in parallel.

There is no difference between mutation objects and query objects. I just treated the top-level mutant objects specially and let them execute them in order.

**Comment by meh** [comment]
Mmh, so I guess using MergedObject would work in this case? Assuming I'm okay with the nested parallelism.

**Comment by sunli829** [comment]
Yes, if you don't care about the order of execution, you can directly use `MergedObject`.

**Comment by sunli829** [comment]
I added the `serial` attribute to the `SimpleObject` and `Object` macros to solve this problem, and now I am going to add it to `MergedObject`.

**Comment by meh** [comment]
Thank you!

**Comment by meh** [comment]
Send help :panda_face: 

```
The application panicked (crashed).
Message:  IndexMap: key not found
Location: /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/model/type.rs:37

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: async_graphql::graphql::field with path=__schema.types.0.fields.5.type.ofType.ofType.ofType parent_type=__Type return_type=__Type
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   1: async_graphql::graphql::field with path=__schema.types.0.fields.5.type.ofType.ofType parent_type=__Type return_type=__Type
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   2: async_graphql::graphql::field with path=__schema.types.0.fields.5.type.ofType parent_type=__Type return_type=__Type
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   3: async_graphql::graphql::field with path=__schema.types.0.fields.5.type parent_type=__Field return_type=__Type!
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   4: async_graphql::graphql::field with path=__schema.types.0.fields.5 parent_type=[__Field!] return_type=__Field!
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   5: async_graphql::graphql::field with path=__schema.types.0.fields parent_type=__Type return_type=[__Field!]
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   6: async_graphql::graphql::field with path=__schema.types.0 parent_type=[__Type!] return_type=__Type!
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   7: async_graphql::graphql::field with path=__schema.types parent_type=__Schema return_type=[__Type!]!
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   8: async_graphql::graphql::field with path=__schema parent_type=Query return_type=__Schema!
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:136
   9: async_graphql::graphql::execute
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:122
  10: async_graphql::graphql::request
      at /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/extensions/tracing.rs:56
  11: query_service::request with request.id=531b1e79-b454-4367-a5ad-41b833315b73 request.remote_addr=Some(127.0.0.1:57122) request.user_agent=Some("Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36") request.method=POST request.path=/
      at query/service/src/lib.rs:167
```

**Comment by sunli829** [comment]
Most likely because you have two objects of different types but the same GraphQL name.

**Comment by meh** [comment]
Damn, how do you suggest I look for this issue? D:

**Comment by sunli829** [comment]
Sorry, this seems to be another problem. Would you please give me an example that can reproduce this problem?  🙂

**Comment by meh** [comment]
Coming up with a reduced test case would be hard, it's already 10k lines of stuff :(

**Comment by sunli829** [comment]
> Coming up with a reduced test case would be hard, it's already 10k lines of stuff :(
> 
> It started happening after I did the MergedObject thing, so it's either something with that, or me starting to use visible = false, which would make sense with the panic during introspection.

Thank you, I am looking for the cause of the problem.

**Comment by sunli829** [comment]
Did you create two objects with the same GraphQL name, like this:

```rust
mod a {
     struct MutationA {}

     #[Object] // <------- name is MutationA
     impl MutationA {}
}

struct MutationA {}

#[Object] // <------- name is MutationA
impl MutationA {}

#[derive(MergedObject)]
struct Mutation(a::MutationA, MutationA);

```

**Comment by meh** [comment]
Not intentionally, I'm explicitly giving names to everything I think.

It is entirely possible I made a mistake there, but I have a ton of types so it's hard to tell with some hint on where the error is.

**Comment by sunli829** [comment]
Look at which key is not found, this is a type name, and then you can find the related type definition based on this.

```
The application panicked (crashed).
Message:  IndexMap: key not found
Location: /home/meh/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-2.9.2/src/model/type.rs:37
```

**Comment by meh** [comment]
Is it in the back trace? Because I didn't notice anything relevant, all I could see was the index of the type it's looking for, but no name.

**Comment by sunli829** [comment]
I added the error output on the master branch, and you can see this value.

**Comment by meh** [comment]
Cool, gonna try in a second, thanks!

**Comment by meh** [comment]
Okay, this makes no sense to me, it's complaining it cannot find type Project, but type Project is definitely there.

**Comment by sunli829** [comment]
Very important, this bug is related to the type definition of `Project`, can you show the type definition code related to `Project`?

**Comment by meh** [comment]
```rust
#[derive(Clone, FromRow, Debug)]
pub struct Project {
	pub id: schema::Id<Project>,
	pub owner: schema::Id<Entity>,

	pub name: Option<String>,

	pub created_at: DateTime<Utc>,
	pub updated_at: Option<DateTime<Utc>>,
	pub deleted_at: Option<DateTime<Utc>>,
}

#[derive(Copy, Clone, Default, Debug)]
pub struct Query;

#[derive(Copy, Clone, Default, Debug)]
pub struct Mutation;

#[derive(Copy, Clone, Debug)]
pub struct MutationWith {
	pub project: schema::Id<Project>,
}

#[derive(Clone, Debug, graphql::MergedObject)]
#[graphql(name = "ProjectMutation")]
pub struct MutationResult(pub Project, pub MutationWith);
```

These are the definitions I've got, before I added the `MutationResult` stuff everything was working fine.

**Comment by meh** [comment]
It's definitely `MergedObject`, the `graphql::Object` is defined in a different module and after that struct is defined, could that be the problem?

**Comment by sunli829** [comment]
I can reproduce this problem now, it is related to `MergedObject`.

**Comment by sunli829** [comment]
@meh Released in `v2.9.4`

**Comment by meh** [comment]
It all works, thank you so much!

**Comment by phungtranminh** [comment]
Hi @sunli829

I am upgrading from `2.8.3` and having some issues with `2.9.4`

I am using `MergedObject` for both `Query` and `Mutation`, not sure what the issue is but it seems like some (or maybe all) mutations are not registered anymore.

Is there any way for me to debug how `async-graphql` is working, or is there any breaking changes you might be aware of?

Happy to open a new issue if this turns out to be a bigger issue.

Thanks,

**Comment by sunli829** [comment]
Please make sure that the GraphQL type name corresponding to each Rust types is different. If this does not solve your current problem, please provide an example that can reproduce this problem.😀

**Comment by phungleson** [comment]
Righto, I changed from this

```
struct MutationRoot(
    a::Mutation,
    b::Mutation,
}
```

to this

```
struct MutationRoot(
    a::AaMutation,
    b::BbMutation,
}
```

And it works again, thanks for quick response!

Checked the book ant it is updated 💯
